### ISSUE TITLE
Support for fractional ctrl-scrolling through stack

### ISSUE THREAD

**Issue by jni** [napari/napari#1514]
## 🚀 Feature

In #1434 we added support for using the scroll-wheel while holding Ctrl to change slices. However, this does not work well with a trackpad rather than a mouse, because the trackpad allows deltas <1, so holding Ctrl and slowly scrolling (ie two finger drag on Mac) actually does nothing but accumulate a whole bunch of events with delta <1. Ideally, we would add an accumulator value that "ticks over" and triggers a scroll when it adds up to more than 1 (or less than -1). The accumulator should be cleared when Ctrl is released, which on the plus side is something that is quite easy to detect! (In earlier discussions we mentioned timers etc which is hairy.)

Here is the relevant bit of code:

https://github.com/napari/napari/blob/936d2890410ac2dc92a42e397bfd7daec37657ea/napari/components/viewer_model.py#L399-L408

Here is an untested start:

https://github.com/napari/napari/pull/1434#discussion_r456730417

**Comment by sofroniewn** [comment]
I just had an idea - given that two finder scrolling when the slider is in focus works nicely to scroll through the stack do we just want holding the `ctrl` key to temporarily toggle focus to the last used slider dims? Curious what @jni @tlambert03 think of this idea?

**Comment by tlambert03** [comment]
yeah, that seems nice... what currently happens if you try two-finger scrolling with the `ctrl` down if no slider has focus?  And... isn't there always a "current" (last used) slider?

**Comment by sofroniewn** [comment]
>  what currently happens if you try two-finger scrolling with the ctrl down if no slider has focus?

Pre-merge of #1434 holding ctrl and hovering over canvas 2 finger scrolling causes pan-zoom just like without ctrl. After #1434 it doesn't, so we'd need to use the #1434 approach of intercepting the scroll wheel call too, but instead of calling the increment dims methods on the model, I guess we'd just reroute the call to the Qt Slider.

> isn't there always a "current" (last used) slider?

I think if there are no sliders present it is `None` and I'm not sure what the default is after launch before any sliders are touched, but those are both easy problems to fix.

**Comment by tlambert03** [comment]
I think I'm missing the point :) what's the actual behavior this would change/fix?  and is the point here specific for 2-finger-scrolling vs mouse-wheel scrolling?

**Comment by sofroniewn** [comment]
Ah sorry, the trackpad scrolling (2-finger scrolling) when the canvas is in focus doesn’t work well after #1434 because of how the delta’s are accumulated - it can’t really get it “to work” on my laptop, maybe it advances a frame or too then suddenly might advance lots. The plan @jni had that @ziyangczi was going to implement was based on accumulating those deltas.

It just occurred to me that an alternative approach might be to change the qt focus to the dims slider and leverage its ability to do nice trackpad scrolling rather than reimplement our own. I’d make the swap for both trackpad/ mouse wheel which flow through the same code pass.

This approach might have downsides associated with changing focus on a key press, but might ultimately mean less complex code on our end, which I think would be good.

@ziyangczi might want to consider both approaches, but curious @tlambert03 with this additional context if one approach jumps out at you?

@jni might also just tell us all that I’ve gotten it completely wrong here too, I’m which case we’ll ignore the idea completely :-)

**Comment by tlambert03** [comment]
> It just occurred to me that an alternative approach might be to change the qt focus to the dims slider and leverage its ability to do nice trackpad scrolling rather than reimplement our own.

ooooh, sorry, I didn't realize this was an alternative to accumulating the deltas.  Yes, seems like it's worth a try!  hard to say ahead of time which approach will feel more natural.

**Comment by jni** [comment]
As much as I like reducing unsolved problems to already-solved problems, I feel like changing the focus with Ctrl is a very big leap — unless I'm misunderstanding, suddenly, all shortcuts with Ctrl will have their focus changed (?), which seems like a minefield of issues waiting to happen.

**Comment by sofroniewn** [comment]
> suddenly, all shortcuts with Ctrl will have their focus changed (?), which seems like a minefield of issues waiting to happen.

Yeah ... I was a little worried about that too - it could just introduce a lot of unexpected coupling. Maybe we can re-route the mouse event to the dims slider without switching focus. How does that sound?

**Comment by jni** [comment]
> Maybe we can re-route the mouse event to the dims slider

*mouse wheel =)

Then, sure! =)

**Comment by sofroniewn** [comment]
> *mouse wheel =)

Yeah, one option to explore is if ctrl is pressed then the mouse wheel event is blocked on the canvas, but reroutes to the dims slider - that seems pretty nice. If that approach doesn't work, we could try the fractional scrolling one.

**Comment by liu-ziyang** [comment]
Before I jump in and implement anything:
I tried the current version without any change, it does work if you scroll hard, so a sensitivity fix may be needed, however, I noticed that the scroll direction is reversed compared to when I scroll on the dim scroller directly, there may be more fixes required to align the scrolling direction.
My personal experience so far is that I like the responsiveness when I scroll directly on the slider bar, it feels smooth and scrolls just the right amount, plus the direction is aligned. So I am also considering reroute mouse wheel to scroll slider as suggested. BUT, I don't have a mouse! (I have 2 mouse but I have USB-A mouse and only USB-C port on my MacBook, good job apple) so I may need to go back and forth between PC and MacBook during dev cycles, then it really isn't a fair comparison in terms of experience. It would be nice to have someone else take this approach and see if this is better

**Comment by sofroniewn** [comment]
> So I am also considering reroute mouse wheel to scroll slider as suggested. BUT, I don't have a mouse! 

The Qt Wheel Event is the same for trackpad and mouse wheel see in https://doc.qt.io/qt-5/qwheelevent.html, so you should be able to implement the event catching and forwarding for the trackpad during development and then only when finished should we test on mouse wheel. Does that make sense?

I don't think not having a mouse wheel will be a problem for implementing this approach as it can be developed/ debugged with the trackpad - and because it is a simple forwarding should actually work the same for trackpad and mouse wheel.

**Comment by liu-ziyang** [comment]
attempted a fix with https://github.com/napari/napari/pull/1525/files, I tested out with trackpad, works on both inverted or uninverted setting on my MacBook to have right/downwards scroll to scroll to right

**Comment by liu-ziyang** [comment]
shall we close this issue?

**Comment by sofroniewn** [comment]
Yes closed by #1525. Putting "closes ..." or "fixes ..." in the PR can do that automatically, see syntax here https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
